File: CS/1315_Sum_of_Nodes_with_Even-Valued_Grandparent.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public int SumEvenGrandparent(TreeNode root) {
        return helper(null, null, root);
    }
    private int helper(TreeNode grandparent, TreeNode parenet, TreeNode current){
        if(current != null){
            if(grandparent!=null){
                    return ((grandparent.val %2 == 0) ? current.val : 0) + helper(parenet, current, current.right) + helper(parenet, current, current.left);
            }
            else{
                return helper(parenet, current, current.right) + helper(parenet, current, current.left);
            }
        }
        else{
            return 0;
        }
    }
}

File: Cpp/1315_Sum_of_Nodes_with_Even-Valued_Grandparent.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int sumEvenGrandparent(TreeNode* root) {
        int result = 0;
        result += helper(-1, -1, root);
        return result;
    }
    int helper(int grandparent, int parent, TreeNode* root){
        if(root == nullptr){
            return 0;
        }
        
        if(grandparent%2 == 0){
            return root->val + helper(parent, root->val, root->left) + helper(parent, root->val, root->right);
        }
        else{
            return helper(parent, root->val, root->left) + helper(parent, root->val, root->right);
        }
    }
};
